### PR TITLE
Globalfunctions seperated from global constraints

### PR DESCRIPTION
**Pull request by Dimosts**
Based on the discussion in #343, I did split the GlobalConstraint and NumericGlobalConstraint classes. I renamed the NumericGlobalConstraint to GlobalFunctions, as this was the main point of splitting them, that they are not really constraints. I have strong feelings about keeping the name GlobalConstraint for the boolean global expressions, as it is a well-established name in CP community.

I think this PR has to be merged before #335 so the necessary changes can be done there to.

### CODE REVIEW

File: cpmpy/expressions/globalfunctions.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    `iv4 + Minimum(iv1, iv2, iv3)`.
+
+    Hence, the CPMpy global functions only capture the `Minimum(iv1, iv2, iv3)` part, whose return type
+    is numeric and can be used in any other CPMpy expression. Only at the time of transforming the CPMpy
+    model to the solver API, will the expressions be decomposed and auxiliary variables introduced as needed
+    such that the solver only receives `Minimum(iv1, iv2, iv3) == ivX` expressions.
+    This is the burden of the CPMpy framework, not of the user who wants to express a problem formulation.
+
+
+    Subclassing GlobalConstraint
```
Comment (IgnaceBleukx) [inline cpmpy/expressions/globalfunctions.py:31]: update this title

File: cpmpy/transformations/decompose_global.py [hunk 2]
```diff
@@ -30,12 +31,14 @@ def decompose_global(lst_of_expr, supported=set(), supported_reif=set()):
 
     """
     def _is_supported(cpm_expr, reified):
-        if isinstance(cpm_expr, GlobalConstraint) and cpm_expr.is_bool():
+        if isinstance(cpm_expr, GlobalConstraint):
+            print("here ? is_supported")
```
Comment (IgnaceBleukx) [inline cpmpy/transformations/decompose_global.py:35]: stray print

File: cpmpy/transformations/decompose_global.py [hunk 3]
```diff
@@ -58,6 +61,7 @@ def _is_supported(cpm_expr, reified):
         decomp_idx = None
 
         if hasattr(cpm_expr, "decompose") and not _is_supported(cpm_expr, reified=False):
+            print("not here?")
```
Comment (IgnaceBleukx) [inline cpmpy/transformations/decompose_global.py:64]: stray print

File: cpmpy/expressions/globalconstraints.py [hunk 4]
```diff
@@ -121,26 +121,20 @@ class GlobalConstraint(Expression):
...
     """
-    # is_bool: whether this is normal constraint (True or False)
-    #   not is_bool: it computes a numeric value (ex: Minimum, Element)
-    def __init__(self, name, arg_list, is_bool=True):
-        super().__init__(name, arg_list)
-        self._is_bool = is_bool
 
     def is_bool(self):
-        """ is it a Boolean (return type) Operator?
+        """ is it a Boolean (return type) Operator? Should always be bool now
```
Comment (IgnaceBleukx) [inline cpmpy/expressions/globalconstraints.py:128]: this doc can be somewhat more professional ;)

File: cpmpy/transformations/decompose_global.py (hunk 2 above)
Comment (Dimosts) [reply cpmpy/transformations/decompose_global.py:35]: woops, forgot the debug message

**Comment by Dimosts** [comment]
It is ready after updating it with the new master (decomposing in tree etc) and can be merged

File: cpmpy/expressions/__init__.py [hunk 5]
```diff
@@ -20,7 +21,8 @@
 # others need to be imported by the developer explicitely
 from .variables import boolvar, intvar, cpm_array
 from .variables import BoolVar, IntVar, cparray # Old, to be deprecated
-from .globalconstraints import AllDifferent, AllDifferentExcept0, AllEqual, Circuit, Inverse, Table, Minimum, Maximum, Element, Xor, Cumulative, IfThenElse, Count, GlobalCardinalityCount, DirectConstraint
+from .globalconstraints import AllDifferent, AllDifferentExcept0, AllEqual, Circuit, Inverse, Table, Xor, Cumulative, IfThenElse, GlobalCardinalityCount, DirectConstraint
```
Comment (JoD) [inline cpmpy/expressions/__init__.py:24]: `GlobalCardinalityCount` is also a GlobalFunction?

Comment (JoD) [reply cpmpy/expressions/__init__.py:24]: My bad, this is not the case.

**Comment by JoD** [comment]
Looks good to me, merged with master, let's see what the online checks say.

**Comment by JoD** [comment]
Allright, let's merge?

**Pull request merged**